### PR TITLE
Fix pattern route performance

### PR DESCRIPTION
**Pull request by gigitux**
PR linked with https://github.com/woocommerce/woocommerce/pull/41168

## What

Fixes #

## Why

This PR splits the `/patterns` endpoint into different endpoints to allow the client to parallelize the requests. The new endpoints are:

`/private/ai/images` -> generate images and return the client the image URLs.
`/private/ai/products` -> generate the payload to pass to /ai/product.
`/private/ai/product` -> update a specific product.
`/private/ai/patterns` -> generate and save patterns.

In the next iteration, we should improve how we handle the errors, but for now, these endpoints are private, so it should not be a big issue. 

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

- Install and activate [this version](https://github.com/woocommerce/woocommerce-blocks/files/13248734/woocommerce.zip) of WooCommerce.
- Install and activate WooCommerce Blocks from the following zip file: [woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13248736/woocommerce-gutenberg-products-block.zip)
- Install and activate Jetpack.
- In your wp-admin, you should see the following section for Jetpack:

Screenshot 2023-10-22 at 09 29 08

- Click on "Set up Jetpack" and make sure the connection is successful:

Screenshot 2023-10-22 at 09 29 08

- Install and activate WooCommerce Beta Tester (the plugin is available [within the monorepo](https://github.com/woocommerce/woocommerce/tree/trunk/plugins))
- Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag:

Screenshot 2023-10-22 at 10 32 24

### Test the AI-managed images

1. Head over to Home > Customize your Store:
 
Screenshot 2023-10-22 at 09 29 08

2. Click on design with AI.
3. Type a description for your business.
4. Make sure the AI content is properly generated and images are assigned to patterns and products based on the description provided.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [ ] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Add suggested changelog entry here.

### CODE REVIEW

File: src/StoreApi/Schemas/V1/AI/BusinessDescriptionSchema.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	 * Patterns schema properties.
+	 *
+	 * @return array
+	 */
+	public function get_properties() {
+		return [];
+	}
+
+	/**
+	 * Get the Patterns response.
```
Comment (albarin) [inline src/StoreApi/Schemas/V1/AI/BusinessDescriptionSchema.php]: I think the `get_item_response` comment is the same in all schemas files.

File: src/StoreApi/Routes/V1/AI/BusinessDescription.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+					],
+				],
+			],
+			'schema'      => [ $this->schema, 'get_public_item_schema' ],
+			'allow_batch' => [ 'v1' => true ],
+		];
+	}
+
+	/**
+	 * Ensure the content and images in patterns are powered by AI.
```
Comment (albarin) [inline src/StoreApi/Routes/V1/AI/BusinessDescription.php]: Same here, it's the same comment for a few of the endpoints.

**Review by albarin — APPROVED** [review]
It's working great, amazing job 👏 
Just left a couple of small comments regarding repeated comments, nothing important.

**Comment by gigitux** [comment]
> Ensure the content and images in patterns are powered by AI.

Thanks for the review! I update the comments with [f96e698](https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/f96e6987cd295dac32ea09a614c6bb518a57c71f).

**Comment by mikejolley** [comment]
Moving to 11.6.0

**Comment by nefeline** [comment]
Great work here @gigitux 🎉 ! Confirmed that with this change, we have a significant boost in performance, well done!

I hope you don't mind I pushed a few changes here to address some issues found during testing and also to remove redundancies:

On https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/827bc3cc8aa4e8e34b74cdd55cf27260e2864655 , the `generate_content` method was updated, and the new `fetch_dummy_products_to_update` method was introduced for handling the fetch of dummy products to be updated. This change was made to ensure dummy products are fetched just once, by the `product` endpoint, and not twice (on both products and product endpoint). The max execution time increase was also removed as now with the requests being parallelized, it is not required anymore.

On https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/7c69cef9c79788b09253dcd9b945c17ee9e8f221 , a change was made to ensure the Product endpoint relies on the newly created `fetch_dummy_products_to_update` method for fetching dummy products to avoid code repetition. On this commit, safety checks were also added and handled errors in case certain properties were not available.

On https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/4a9ff4a5bc6d41add247b4e356141fd0c8d4fae6 a change was made to ensure the `$populate_products` variable is not a WP_Error _before_ trying to access its properties. Also added an extra safety check for the existence of the `$populate_products['product_content']` attribute.

On https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/734e48569019552a10c17b80ece9bd50607b6420. The memory increase was also removed and doc blocks updated. 

I can confirm the feature works, but it would be great if more folks could re-test the PR with the latest changes to ensure everything is working as expected. cc: @albarin @tjcafferkey. Here's the latest zip file for testing: 
[woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13277522/woocommerce-gutenberg-products-block.zip)

**Comment by tjcafferkey** [comment]
When I generate a store using the experience I land on the following page:

Screenshot 2023-11-07 at 09 58 46

However, when I click on the homepage templates I get correctly generated templates and patterns as seen below.

Screenshot 2023-11-07 at 09 58 51

Is this a known issue at this point, and thus expected?

**Comment by tjcafferkey** [comment]
I also get inaccurate text, pretty sure this is a known issue with this pattern but raising for confirmation?

Screenshot 2023-11-07 at 10 01 00

**Comment by gigitux** [comment]
Thanks for the review, @tjcafferkey!

> When I generate a store using the experience I land on the following page:

This issue https://github.com/woocommerce/woocommerce/issues/41229 was fixed yesterday and the Core build uploaded here doesn't contain the patch.

> I also get inaccurate text, pretty sure this is a known issue with this pattern but raising for confirmation?

Just to confirm, for "inaccurate text" are you referring to the duplicate text? If yes, @albarin fixed it (https://github.com/woocommerce/woocommerce-blocks/pull/11612), but this branch doesn't contain the patch.

**Comment by tjcafferkey** [comment]
Thanks @gigitux I will continue to review the code now.

**Review by tjcafferkey — APPROVED** [review]

**Comment by nefeline** [comment]
@gigitux @tjcafferkey FYI: I just noticed that we could occasionally have products with content not updated without increasing the max_execution_time, so I went ahead and re-added it here since it is still required: https://github.com/woocommerce/woocommerce-blocks/pull/11535/commits/31daa3332128bbb325512c101d0c37d06b68ce13

**Pull request merged**